Extensible jacobian and network jacobian interfaces

anthony-walker opened the pull request:
<!-- Thanks for contributing code! Please include a description of your change and check your pull request against the list below. For further questions, refer to the contributing guide (https://github.com/Cantera/cantera/blob/main/CONTRIBUTING.md). -->

**Changes proposed in this pull request**

In this pull request I have setup basic structure for adding contributions to the jacobian from flow devices and walls.
I have also added an extensible jacobian interface to the `ExtensibleReactor` system and modified the existing jacobian system to pass a vector of triplets as is needed in lieu of creating a large number of separate sparse matrices. 

@speth @ischoegl some additional tests are needed for the wall and flow contributions but I wanted to get the content officially into a PR.

<!-- Provide a clear and concise description of changes and/or features introduced in this pull request. -->

- Jacobian contributions from walls and flow devices
- Extensible jacobian interface
- Enhancements to existing jacobian backend

**Checklist**

- [x] The pull request includes a clear description of this code change
- [x] Commit messages have short titles and reference relevant issues
- [x] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [x] Style & formatting of contributed code follows [contributing guidelines](https://github.com/Cantera/cantera/blob/main/CONTRIBUTING.md)
- [x] The pull request is ready for review

anthony-walker committed:
Extend jacobian building to Extensible Reactor system

anthony-walker committed:
Complete finite difference jacobian

anthony-walker committed:
Build jacobian structure defined for walls and flow devices

anthony-walker committed:
Update Extensible Jacobian interface to not reconstruct entire triplet vector

anthony-walker committed:
Cpp Wall Jacobian Test

anthony-walker committed:
Doxygen fixes

anthony-walker committed:
Add tests, fix bugs, remove added unused code in cython and cpp

ischoegl reviewed (commented):
While I don't have enough experience in extensible things for a complete review, I noticed that a lot of test are failing with:
```
*******************************************************************************
NotImplementedError thrown by MassFlowController::buildReactorJacobian:
Not implemented.
*******************************************************************************
```
which doesn't appear to be .NET related.

anthony-walker commented:
@ischoegl these failures are from my latest bug fixes there is a separate commit which addresses the `.NET` failure issue.

speth commented:
Before you get any further, I just wanted to suggest undoing the merge commit and rebasing onto the current `main` branch instead.

anthony-walker committed:
Add a default evaluation function for reactors

It can sometimes be useful to add on to the default evaluation in such a
way that the standard before or after delegations cannot handle. A default
evaluation function allows for this.

anthony-walker committed:
Working on PR updates, have to fix a few more bugs
in regards to equations, handling volumes, etc.

speth reviewed (changes requested):
Thanks for putting this together, @anthony-walker. I think it helps in laying out the necessary pieces for handling the cross-reactor terms in the Jacobian -- in particular, the mechanics surrounding `Wall::buildNetworkJacobian` seem like a good basis to keep building on.

I think there is some room for simplification of this structure, which I've left specific comments on. However, I think there is a significant issue with sorting out exactly which derivatives need to be computed and what the right formulas are for the different cases. The challenge I see us needing to resolve to generalize this across the different reactor types is that what derivative is needed for a single interaction (for example, wall heat transfer) depends on *both* reactor types and what variables they use to represent each term, and we need a way of selecting the correct formula for each Jacobian element based on that pair of reactor types.

speth commented:
For an example of where the current derivatives are clearly not correct, consider the following example, adapted from one of your new tests:
```py
import cantera as ct
import numpy as np

# create first reactor
gas1 = ct.Solution("h2o2.yaml", "ohmech")
gas1.TPX = 600, ct.one_atm, "O2:1.0"
r1 = ct.IdealGasMoleReactor(gas1)
r1.name = 'R1'
gas1.set_multiplier(0.0)

# create second reactor
gas2 = ct.Solution("h2o2.yaml", "ohmech")
gas2.TPX = 300, ct.one_atm, "O2:1.0"
gas2.set_multiplier(0.0)
r2 = ct.IdealGasMoleReactor(gas2)
r2.name = 'R2'

# create wall
U = 500.0
A = 3.0
w = ct.Wall(r1, r2, U=U, A=A)
net = ct.ReactorNet([r1, r2])

def print_jac(J):
    for i,j in zip(*np.nonzero(J)):
        name_i = net.component_name(i).replace('temperature', 'T')
        name_j = net.component_name(j).replace('temperature', 'T')
        net.component_name(i).replace('temperature', 'T')
        #print(f"J[{i: 3d}, {j: 3d}] = {J[i,j]:.3e}")
        print(f"J[{name_i:8s}, {name_j:8s}] = {J[i,j]:.3e}")

print('Semi-analytical Jacobian:')
print_jac(net.jacobian)
print('\nFinite difference Jacobian:')
print_jac(net.finite_difference_jacobian)
```
which outputs:
```
Semi-analytical Jacobian:
J[R1: T   , R1: T   ] = -2.727e+00
J[R1: T   , R1: H2  ] = 9.000e+05
J[R1: T   , R1: H   ] = 9.000e+05
J[R1: T   , R1: O   ] = 9.000e+05
J[R1: T   , R1: O2  ] = 9.000e+05
J[R1: T   , R1: OH  ] = 9.000e+05
J[R1: T   , R1: H2O ] = 9.000e+05
J[R1: T   , R1: HO2 ] = 9.000e+05
J[R1: T   , R1: H2O2] = 9.000e+05
J[R1: T   , R1: AR  ] = 9.000e+05
J[R1: T   , R1: N2  ] = 9.000e+05
J[R1: T   , R2: H2  ] = -4.500e+05
J[R1: T   , R2: H   ] = -4.500e+05
J[R1: T   , R2: O   ] = -4.500e+05
J[R1: T   , R2: O2  ] = -4.500e+05
J[R1: T   , R2: OH  ] = -4.500e+05
J[R1: T   , R2: H2O ] = -4.500e+05
J[R1: T   , R2: HO2 ] = -4.500e+05
J[R1: T   , R2: H2O2] = -4.500e+05
J[R1: T   , R2: AR  ] = -4.500e+05
J[R1: T   , R2: N2  ] = -4.500e+05
J[R2: T   , R1: H2  ] = 9.000e+05
J[R2: T   , R1: H   ] = 9.000e+05
J[R2: T   , R1: O   ] = 9.000e+05
J[R2: T   , R1: O2  ] = 9.000e+05
J[R2: T   , R1: OH  ] = 9.000e+05
J[R2: T   , R1: H2O ] = 9.000e+05
J[R2: T   , R1: HO2 ] = 9.000e+05
J[R2: T   , R1: H2O2] = 9.000e+05
J[R2: T   , R1: AR  ] = 9.000e+05
J[R2: T   , R1: N2  ] = 9.000e+05
J[R2: T   , R2: T   ] = -1.887e+00
J[R2: T   , R2: H2  ] = -4.500e+05
J[R2: T   , R2: H   ] = -4.500e+05
J[R2: T   , R2: O   ] = -4.500e+05
J[R2: T   , R2: O2  ] = -4.500e+05
J[R2: T   , R2: OH  ] = -4.500e+05
J[R2: T   , R2: H2O ] = -4.500e+05
J[R2: T   , R2: HO2 ] = -4.500e+05
J[R2: T   , R2: H2O2] = -4.500e+05
J[R2: T   , R2: AR  ] = -4.500e+05
J[R2: T   , R2: N2  ] = -4.500e+05

Finite difference Jacobian:
J[R1: T   , R1: T   ] = -2.727e+00
J[R1: T   , R1: O2  ] = 4.589e+04
J[R1: T   , R2: T   ] = 3.107e+00
J[R2: T   , R1: T   ] = 1.752e+00
J[R2: T   , R2: T   ] = -1.887e+00
J[R2: T   , R2: O2  ] = -1.294e+04
```

anthony-walker commented on test/python/conftest.py:
```diff
@@ -1,3 +1,26 @@
 
 pytest.register_assert_rewrite("pint.testing")
+
+def pytest_addoption(parser):
+    # this option is meant for debugging usage only
+    parser.addoption(
+        "--diagnose", action="store_true",
+        default=False,
+        help="Run only diagnostic tests, debugging purposes only.")
+
+
+def pytest_configure(config):
+    config.addinivalue_line("markers",
+                            "diagnose: mark to only run this test.")
+
+
+def pytest_collection_modifyitems(config, items):
+    # run only diagnostic tests
+    if config.getoption("--diagnose"):
+        reason = "running only diagnostic tests."
+        skip_diagnose = pytest.mark.skip(reason=reason)
+        for item in items:
+            if "diagnose" not in item.keywords:
+                item.add_marker(skip_diagnose)
+        return
```
This is to isolate python tests. I will add to docs.

anthony-walker commented on src/zeroD/Wall.cpp:
```diff
@@ -111,4 +113,66 @@ double Wall::heatRate()
+    size_t l_net = m_left->network().globalStartIndex((Reactor* ) m_left);
+    size_t l_eidx = m_left->energyIndex();
+
+    if (((Reactor* ) m_right)->energyEnabled()) {
```
@speth I think it would make sense to move chemistryEnabled in the same way for consistency.

anthony-walker commented on src/zeroD/Reactor.cpp:
```diff
@@ -37,6 +37,17 @@ void Reactor::setDerivativeSettings(AnyMap& settings)
+
+    // set reactor settings
+    bool force = settings.empty();
+    if (force || settings.hasKey("skip-walls")) {
+        m_jac_skip_walls = settings.getBool("skip-walls",
+            false);
+    }
+    if (force || settings.hasKey("skip-flow-devices")) {
+        m_jac_skip_flow_devices = settings.getBool("skip-flow-devices",
+            false);
+    }
```
That is a good question, I believe depending on the flow device there could be performance implications.

anthony-walker commented on include/cantera/zeroD/Reactor.h:
```diff
@@ -212,9 +212,50 @@ class Reactor : public ReactorBase
+        throw NotImplementedError(type() + "::buildJacobian");
     }
 
+    //! Calculate the Jacobian of a Reactor specialization for wall contributions.
+    //! @param jacVector vector where jacobian triplets are added
+    //! @warning Depending on the particular implementation, this may return an
+    //! approximate Jacobian intended only for use in forming a preconditioner for
+    //! iterative solvers.
+    //! @ingroup derivGroup
+    //!
+    //! @warning  This method is an experimental part of the %Cantera
+    //! API and may be changed or removed without notice.
+    virtual void buildWallJacobian(vector<Eigen::Triplet<double>>& jacVector);
+
+    //! Calculate flow contributions to the Jacobian of a Reactor specialization.
+    //! @param jacVector vector where jacobian triplets are added
+    //! @warning Depending on the particular implementation, this may return an
+    //! approximate Jacobian intended only for use in forming a preconditioner for
+    //! iterative solvers.
+    //! @ingroup derivGroup
+    //!
+    //! @warning  This method is an experimental part of the %Cantera
+    //! API and may be changed or removed without notice.
+    virtual void buildFlowJacobian(vector<Eigen::Triplet<double>>& jacVector);
```
My thought was when an individual wants to access a reactor only Jacobian.

anthony-walker commented on include/cantera/zeroD/FlowDevice.h:
```diff
@@ -114,7 +115,56 @@ class FlowDevice
+    //! or removed without notice.
+    //! @since New in %Cantera 3.0.
+    //!
+    virtual void buildNetworkJacobian(vector<Eigen::Triplet<double>>& jacVector) {
+        if (!m_jac_calculated) {
+            throw NotImplementedError(type() + "::buildNetworkJacobian");
+        }
+    }
```
There is not one currently.

speth committed:
[Test] Replace deprecated syntax in reactor tests

speth committed:
[Python] Fix derived pointer initialization for Jacobians

speth committed:
[Reactor] Fix species counting in case of user-defined solution components

speth committed:
[Reactor] Fix duplicate calculation of connector Jacobians

ischoegl commented:
Superseded by #2113.
